micro-rdk/src/esp32/coredump.rs

```diff
@@ -0,0 +1,196 @@
+        let res = GenericReadingsResult::from([(
+            "has_coredump".to_owned(),
+            Value {
+                kind: Some(Kind::BoolValue(has_coredump)),
+            },
+        )]);
+        Ok(res)
+    }
+}
+
+impl Status for Coredump {
+    fn get_status(
+        &self,
+    ) -> Result<Option<crate::google::protobuf::Struct>, crate::common::status::StatusError> {
+        Ok(Some(protobuf::Struct {
+            fields: HashMap::new(),
+        }))
+    }
+}
+
+struct Null;
+
+const CORE_FRAGMENT_SIZE: usize = 3072;
+
+impl DoCommand for Coredump {
+    fn do_command(
+        &mut self,
+        command_struct: Option<protobuf::Struct>,
+    ) -> Result<Option<protobuf::Struct>, crate::common::generic::GenericError> {
+        if let Some(cmd) = command_struct {
+            // Coredump would be chunked into 4096 bytes encoded in Base64
+            // len(bases64) = 4*(n/3) therefore n = (4096/4)*3 rounded down
+            // which yield n = 3072
+            if cmd.fields.get("sizes").is_some() {
+                return Ok(Some(Struct {
+                    fields: HashMap::from([
+                        (
+                            "nb_chunk".to_owned(),
+                            Value {
+                                kind: Some(Kind::NumberValue(
+                                    (self.len / CORE_FRAGMENT_SIZE) as f64,
+                                )),
+                            },
+                        ),
+                        (
+                            "len".to_owned(),
+                            Value {
+                                kind: Some(Kind::NumberValue(self.len as f64)),
+                            },
+                        ),
+                    ]),
+                }));
+            }
+            if cmd.fields.get("erase_coredump").is_some() {
+                esp_idf_svc::sys::esp!(unsafe {
+                    esp_partition_erase_range(
+                        self.coredump_partition_ptr,
+                        0_usize,
+                        self.coredump_partition_ptr.size as usize,
+                    )
+                })
+                .map_err(|e| GenericError::Other(e.into()))?;
+                self.len = 0;
+                return Ok(None);
+            }
+            if let Some(Kind::NumberValue(val)) = cmd
+                .fields
+                .get("get_nth_chunk")
+                .and_then(|v| v.kind.as_ref())
+            {
+                let offset = ((val.floor() as usize) * CORE_FRAGMENT_SIZE).min(self.len);
+                offset
+                    .lt(&self.len)
+                    .then_some(Null)
+                    .ok_or(GenericError::Other(
+                        "chunk requested outside of bounds".into(),
+                    ))?;
+
+                let to_read: usize = CORE_FRAGMENT_SIZE.min(self.len - offset);
+                let mut buf = BytesMut::with_capacity(CORE_FRAGMENT_SIZE);
```

Comment:
I'm not sure if I'm reading this right. is this 0-indexed?

`offset`: the smaller of `the starting byte of the n'th chunk` or the `len` of coredump 

if `offset < self.len`, then turn to `Option<Null>`, converted to a Result<Null, Err>, mapping on the Err as out of bounds.

Wouldn't you want L152 (after getting offset) to compare `(offset + CORE_FRAGMENT_SIZE).lt(&self.len)` to ensure that the end of the nth chunk is within boundaries, not just its starting point?

Reply:
yes it's 0 indexed

offset is the lower bound, to_read+offset will give you the upper bound. Because we will read [offset, offset + min(CORE_FRAGMENT_SIZE, self.len -offset))